mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -80,69 +50,157 @@ class LogisticRegression extends Estimator[LogisticRegressionModel] with Logisti
+    1.0 / (1.0 + math.exp(-m))
   }
 
   override def transform(dataset: SchemaRDD, paramMap: ParamMap): SchemaRDD = {
+    // Check schema
     transformSchema(dataset.schema, paramMap, logging = true)
+
     import dataset.sqlContext._
     val map = this.paramMap ++ paramMap
-    val score: Vector => Double = (v) => {
-      val margin = BLAS.dot(v, weights)
-      1.0 / (1.0 + math.exp(-margin))
+
+    // Output selected columns only.
+    // This is a bit complicated since it tries to avoid repeated computation.
+    //   rawPrediction (-margin, margin)
+    //   probability (1.0-score, score)
+    //   prediction (max margin)
+    var tmpData = dataset
+    var numColsOutput = 0
+    if (map(rawPredictionCol) != "") {
+      val features2raw: Vector => Vector = predictRaw
+      tmpData = tmpData.select(Star(None),
+        features2raw.call(map(featuresCol).attr) as map(rawPredictionCol))
+      numColsOutput += 1
+    }
+    if (map(probabilityCol) != "") {
+      if (map(rawPredictionCol) != "") {
+        val raw2prob: Vector => Vector = (rawPreds) => {
+          val prob1 = 1.0 / (1.0 + math.exp(-rawPreds(1)))
+          Vectors.dense(1.0 - prob1, prob1)
+        }
+        tmpData = tmpData.select(Star(None),
+          raw2prob.call(map(rawPredictionCol).attr) as map(probabilityCol))
+      } else {
+        val features2prob: Vector => Vector = predictProbabilities
+        tmpData = tmpData.select(Star(None),
+          features2prob.call(map(featuresCol).attr) as map(probabilityCol))
+      }
+      numColsOutput += 1
+    }
+    if (map(predictionCol) != "") {
+      val t = map(threshold)
+      if (map(probabilityCol) != "") {
+        val predict: Vector => Double = (probs) => {
+          if (probs(1) > t) 1.0 else 0.0
+        }
+        tmpData = tmpData.select(Star(None),
+          predict.call(map(probabilityCol).attr) as map(predictionCol))
+      } else if (map(rawPredictionCol) != "") {
+        val predict: Vector => Double = (rawPreds) => {
+          val prob1 = 1.0 / (1.0 + math.exp(-rawPreds(1)))
+          if (prob1 > t) 1.0 else 0.0
+        }
+        tmpData = tmpData.select(Star(None),
+          predict.call(map(rawPredictionCol).attr) as map(predictionCol))
+      } else {
+        val predict: Vector => Double = this.predict
+        tmpData = tmpData.select(Star(None),
+          predict.call(map(featuresCol).attr) as map(predictionCol))
+      }
+      numColsOutput += 1
     }
-    val t = map(threshold)
-    val predict: Double => Double = (score) => {
-      if (score > t) 1.0 else 0.0
+    if (numColsOutput == 0) {
+      this.logWarning(s"$uid: LogisticRegressionModel.transform() was called as NOOP" +
+        " since no output columns were set.")
     }
-    dataset.select(Star(None), score.call(map(featuresCol).attr) as map(scoreCol))
-      .select(Star(None), predict.call(map(scoreCol).attr) as map(predictionCol))
+    tmpData
+  }
+
+  override val numClasses: Int = 2
```

Comment:
As with the Enumeration suggestion above - perhaps some enum of size 2 is appropriate here.